File: ios/SpicyPlayeriOS/Sources/TTMLLyricsParser.swift
import Foundation

enum TTMLLyricsParser {
    static func parse(data: Data) throws -> ParsedLyrics {
        let delegate = ParserDelegate()
        let parser = XMLParser(data: data)
        parser.shouldProcessNamespaces = true
        parser.delegate = delegate

        guard parser.parse() else {
            throw delegate.error ?? parser.parserError ?? NSError(
                domain: "SpicyPlayeriOS.TTMLParser",
                code: 1,
                userInfo: [NSLocalizedDescriptionKey: "Failed to parse TTML lyrics."]
            )
        }

        return delegate.result()
    }
}

private final class ParserDelegate: NSObject, XMLParserDelegate {
    struct SpanContext {
        let begin: Int?
        let end: Int?
        let isBackground: Bool
    }

    var lines: [LyricLine] = []
    var songwriters: [String] = []
    var error: Error?

    private var defaultAgent: String?
    private var inSongwriter = false
    private var currentParagraph: ParagraphState?

    func result() -> ParsedLyrics {
        var allLines = lines

        if !songwriters.isEmpty {
            let lastLineEnd = allLines.map(\.endMs).max() ?? 0
            let text = "Written by \(songwriters.joined(separator: ", "))"
            let tokens = text.split(separator: " ").map(String.init)
            let words = tokens.map { token in
                LyricWord(
                    text: token,
                    startMs: lastLineEnd,
                    endMs: lastLineEnd + 1000,
                    isPartOfWord: false,
                    isLetterGroup: false,
                    letters: []
                )
            }

            allLines.append(
                LyricLine(
                    words: words,
                    startMs: lastLineEnd,
                    agent: nil,
                    isBackground: false,
                    oppositeAligned: false,
                    isSongwriter: true,
                    isInterlude: false,
                    interludeEndMs: -1
                )
            )
        }

        let mainLines = allLines
            .filter { !$0.isBackground && !$0.isSongwriter }
            .sorted { $0.startMs < $1.startMs }

        var interludes: [LyricLine] = []
        if let first = mainLines.first, first.startMs >= 3000 {
            interludes.append(
                LyricLine(
                    words: [],
                    startMs: 0,
                    agent: nil,
                    isBackground: false,
                    oppositeAligned: false,
                    isSongwriter: false,
                    isInterlude: true,
                    interludeEndMs: first.startMs
                )
            )
        }

        if mainLines.count >= 2 {
            for index in 0..<(mainLines.count - 1) {
                let gapStart = mainLines[index].endMs
                let gapEnd = mainLines[index + 1].startMs
                if gapEnd - gapStart >= 3000 {
                    interludes.append(
                        LyricLine(
                            words: [],
                            startMs: gapStart,
                            agent: nil,
                            isBackground: false,
                            oppositeAligned: false,
                            isSongwriter: false,
                            isInterlude: true,
                            interludeEndMs: gapEnd
                        )
                    )
                }
            }
        }

        allLines.append(contentsOf: interludes)
        allLines.sort { $0.startMs < $1.startMs }

        return ParsedLyrics(lines: allLines, songwriters: songwriters)
    }

    func parser(_ parser: XMLParser, parseErrorOccurred parseError: Error) {
        error = parseError
    }

    func parser(
        _ parser: XMLParser,
        didStartElement elementName: String,
        namespaceURI: String?,
        qualifiedName qName: String?,
        attributes attributeDict: [String: String] = [:]
    ) {
        let name = normalizedName(elementName, qName)

        switch name {
        case "songwriter":
            inSongwriter = true
        case "agent":
            let identifier = attributeDict["xml:id"] ?? attributeDict["id"]
            if identifier == "v1" {
                defaultAgent = identifier
            }
        case "p":
            let agent = attributeDict["agent"] ?? attributeDict["ttm:agent"]
            if defaultAgent == nil, let agent {
                defaultAgent = agent
            }
            currentParagraph = ParagraphState(
                beginMs: parseTimeMs(attributeDict["begin"]) ?? 0,
                endMs: parseTimeMs(attributeDict["end"]) ?? 0,
                agent: agent,
                defaultAgent: defaultAgent
            )
        case "span":
            currentParagraph?.pushSpan(attributes: attributeDict)
        default:
            break
        }
    }

    func parser(_ parser: XMLParser, foundCharacters string: String) {
        if inSongwriter {
            let trimmed = string.trimmingCharacters(in: .whitespacesAndNewlines)
            if !trimmed.isEmpty {
                songwriters.append(trimmed)
            }
            return
        }

        currentParagraph?.consume(text: string)
    }

    func parser(
        _ parser: XMLParser,
        didEndElement elementName: String,
        namespaceURI: String?,
        qualifiedName qName: String?
    ) {
        let name = normalizedName(elementName, qName)

        switch name {
        case "songwriter":
            inSongwriter = false
        case "span":
            currentParagraph?.popSpan()
        case "p":
            if let paragraph = currentParagraph {
                lines.append(contentsOf: paragraph.makeLines())
            }
            currentParagraph = nil
        default:
            break
        }
    }

    private func normalizedName(_ elementName: String, _ qName: String?) -> String {
        let raw = qName ?? elementName
        return raw.split(separator: ":").last.map(String.init) ?? raw
    }
}

private struct ParagraphState {
    let beginMs: Int
    let endMs: Int
    let agent: String?
    let defaultAgent: String?

    private(set) var leadWords: [LyricWord] = []
    private(set) var backgroundGroups: [[LyricWord]] = []
    private var stack: [ParserDelegate.SpanContext]
    private var previousEndedMidWord = false
    private var currentBackgroundGroup: [LyricWord]?
    private var inBackgroundSpan = false
    private var backgroundPreviousEndedMidWord = false

    init(beginMs: Int, endMs: Int, agent: String?, defaultAgent: String?) {
        self.beginMs = beginMs
        self.endMs = endMs
        self.agent = agent
        self.defaultAgent = defaultAgent
        self.stack = [ParserDelegate.SpanContext(begin: beginMs, end: endMs, isBackground: false)]
    }

    mutating func pushSpan(attributes: [String: String]) {
        let inherited = stack.last ?? ParserDelegate.SpanContext(begin: beginMs, end: endMs, isBackground: false)
        let begin = parseTimeMs(attributes["begin"]) ?? inherited.begin
        let end = parseTimeMs(attributes["end"]) ?? inherited.end
        let role = attributes["role"] ?? attributes["ttm:role"]
        let isBackground = role == "x-bg" || inherited.isBackground

        if isBackground && currentBackgroundGroup == nil {
            currentBackgroundGroup = []
            backgroundPreviousEndedMidWord = false
        }

        inBackgroundSpan = isBackground
        stack.append(ParserDelegate.SpanContext(begin: begin, end: end, isBackground: isBackground))
    }

    mutating func popSpan() {
        guard !stack.isEmpty else {
            return
        }

        let popped = stack.removeLast()
        let nextIsBackground = stack.last?.isBackground ?? false
        if popped.isBackground && !nextIsBackground {
            if let currentBackgroundGroup, !currentBackgroundGroup.isEmpty {
                backgroundGroups.append(currentBackgroundGroup)
            }
            self.currentBackgroundGroup = nil
            inBackgroundSpan = false
        } else {
            inBackgroundSpan = nextIsBackground
        }
    }

    mutating func consume(text rawText: String) {
        let context = stack.last ?? ParserDelegate.SpanContext(begin: beginMs, end: endMs, isBackground: false)
        let isBackgroundToken = context.isBackground || inBackgroundSpan

        if rawText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            if isBackgroundToken {
                backgroundPreviousEndedMidWord = false
            } else {
                previousEndedMidWord = false
            }
            return
        }

        let startsWithSpace = rawText.first?.isWhitespace == true
        let endsWithSpace = rawText.last?.isWhitespace == true
        var trimmed = rawText.trimmingCharacters(in: .whitespacesAndNewlines)

        if isBackgroundToken {
            trimmed = trimmed.replacingOccurrences(of: #"^\("#, with: "", options: .regularExpression)
            trimmed = trimmed.replacingOccurrences(of: #"\)$"#, with: "", options: .regularExpression)
        }

        if trimmed.isEmpty {
            return
        }

        let spaceTokens = trimmed
            .split(whereSeparator: { $0.isWhitespace })
            .map(String.init)
            .filter { !$0.isEmpty }

        var wordsToAdd: [(String, Bool)] = []
        for (tokenIndex, token) in spaceTokens.enumerated() {
            let subTokens = splitKeepingTrailingHyphen(token)
            for (subIndex, subToken) in subTokens.enumerated() {
                let isAttached: Bool
                if tokenIndex == 0 && subIndex == 0 {
                    if isBackgroundToken {
                        isAttached = backgroundPreviousEndedMidWord && !startsWithSpace
                    } else {
                        isAttached = previousEndedMidWord && !startsWithSpace
                    }
                } else {
                    isAttached = subIndex > 0
                }
                wordsToAdd.append((subToken, isAttached))
            }
        }

        let start = context.begin ?? beginMs
        let end = context.end ?? (start + 1000)
        let duration = max(end - start, 0)
        let chunkDuration = wordsToAdd.isEmpty ? duration : duration / wordsToAdd.count

        for (index, entry) in wordsToAdd.enumerated() {
            let wordStart = start + (index * chunkDuration)
            let wordEnd = start + ((index + 1) * chunkDuration)
            let wordDuration = wordEnd - wordStart
            let token = entry.0
            let isLetterGroup = wordDuration >= 1000 && token.count > 1 && token.count <= 12

            let letters: [LyricLetter]
            if isLetterGroup {
                let count = max(token.count, 1)
                let letterDuration = Double(wordDuration) / Double(count)
                letters = token.enumerated().map { offset, character in
                    LyricLetter(
                        char: String(character),
                        startMs: wordStart + Int(Double(offset) * letterDuration),
                        endMs: wordStart + Int(Double(offset + 1) * letterDuration)
                    )
                }
            } else {
                letters = []
            }

            let word = LyricWord(
                text: token,
                startMs: wordStart,
                endMs: wordEnd,
                isPartOfWord: entry.1,
                isLetterGroup: isLetterGroup,
                letters: letters
            )

            if isBackgroundToken {
                currentBackgroundGroup?.append(word)
            } else {
                leadWords.append(word)
            }
        }

        if isBackgroundToken {
            backgroundPreviousEndedMidWord = !endsWithSpace
        } else {
            previousEndedMidWord = !endsWithSpace
        }
    }

    func makeLines() -> [LyricLine] {
        let oppositeAligned = agent != nil && defaultAgent != nil && agent != defaultAgent
        var result: [LyricLine] = []

        result.append(
            LyricLine(
                words: leadWords,
                startMs: beginMs,
                agent: agent,
                isBackground: false,
                oppositeAligned: oppositeAligned,
                isSongwriter: false,
                isInterlude: false,
                interludeEndMs: -1
            )
        )

        for group in backgroundGroups {
            let groupStart = group.first?.startMs ?? beginMs
            result.append(
                LyricLine(
                    words: group,
                    startMs: groupStart,
                    agent: agent,
                    isBackground: true,
                    oppositeAligned: oppositeAligned,
                    isSongwriter: false,
                    isInterlude: false,
                    interludeEndMs: -1
                )
            )
        }

        return result
    }

    private func splitKeepingTrailingHyphen(_ token: String) -> [String] {
        guard token.contains("-") else {
            return [token]
        }

        var result: [String] = []
        var current = ""
        for character in token {
            current.append(character)
            if character == "-" {
                result.append(current)
                current = ""
            }
        }

        if !current.isEmpty {
            result.append(current)
        }

        return result.filter { !$0.isEmpty }
    }
}

private func parseTimeMs(_ time: String?) -> Int? {
    guard let time else {
        return nil
    }

    let parts = time.split(separator: ":").map(String.init)
    if parts.count == 2 {
        let minutes = Int(parts[0]) ?? 0
        let secondsParts = parts[1].split(separator: ".", maxSplits: 1).map(String.init)
        let seconds = Int(secondsParts[0]) ?? 0
        let milliseconds = secondsParts.count > 1 ? paddedMilliseconds(secondsParts[1]) : 0
        return ((minutes * 60) + seconds) * 1000 + milliseconds
    }

    if parts.count == 3 {
        let hours = Int(parts[0]) ?? 0
        let minutes = Int(parts[1]) ?? 0
        let secondsParts = parts[2].split(separator: ".", maxSplits: 1).map(String.init)
        let seconds = Int(secondsParts[0]) ?? 0
        let milliseconds = secondsParts.count > 1 ? paddedMilliseconds(secondsParts[1]) : 0
        return ((hours * 3600) + (minutes * 60) + seconds) * 1000 + milliseconds
    }

    return nil
}

private func paddedMilliseconds(_ raw: String) -> Int {
    let prefix = String(raw.prefix(3))
    let padded = prefix.padding(toLength: 3, withPad: "0", startingAt: 0)
    return Int(padded) ?? 0
}

File: app/src/main/java/com/tx24/spicyplayer/parser/TtmlLyricsParser.kt
package com.tx24.spicyplayer.parser

import android.util.Xml
import com.tx24.spicyplayer.models.Letter
import com.tx24.spicyplayer.models.Line
import com.tx24.spicyplayer.models.ParsedLyrics
import com.tx24.spicyplayer.models.Word
import org.xmlpull.v1.XmlPullParser
import java.io.InputStream
import java.io.InputStreamReader

/**
 * A parser for TTML (Timed Text Markup Language) lyric files.
 * This parser extracts lyrics, timing information, songwriter metadata, and interludes.
 */
object TtmlLyricsParser {
    /**
     * Context used during parsing to keep track of nested span timing and metadata.
     */
    data class SpanContext(
        val begin: Long?,
        val end: Long?,
        val isBg: Boolean = false,
    )

    /**
     * Parses a TTML input stream into a [ParsedLyrics] object.
     *
     * @param inputStream The stream containing the TTML content.
     * @return A [ParsedLyrics] object containing the parsed lines and metadata.
     */
    fun parse(inputStream: InputStream): ParsedLyrics {
        val parser = Xml.newPullParser()
        parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, true)
        val reader = inputStream.reader(Charsets.UTF_8)
        parser.setInput(reader)

        val lines = mutableListOf<Line>()
        val songwriters = mutableListOf<String>()
        var eventType = parser.eventType
        var defaultAgent: String? = null

        var inSongwriters = false

        // Main parsing loop.
        while (eventType != XmlPullParser.END_DOCUMENT) {
            when (eventType) {
                XmlPullParser.START_TAG -> {
                    when (parser.name) {
                        "songwriter" -> inSongwriters = true
                        "p" -> {
                            // Paragraph tags represent a block of lyrics.
                            val agent = parser.getAttributeValue(null, "agent")
                                ?: parser.getAttributeValue("http://www.w3.org/ns/ttml#metadata", "agent")
                            if (defaultAgent == null && agent != null) {
                                defaultAgent = agent
                            }
                            // Parse the paragraph into one or more Line objects.
                            val parsedLines = parseParagraph(parser, agent, defaultAgent)
                            lines.addAll(parsedLines)
                        }
                        "agent" -> {
                            val id = parser.getAttributeValue("http://www.w3.org/XML/1998/namespace", "id")
                                ?: parser.getAttributeValue(null, "id")
                            if (id == "v1") {
                                defaultAgent = id
                            }
                        }
                    }
                }
                XmlPullParser.TEXT -> {
                    // Extract songwriter names from the <songwriter> tag.
                    if (inSongwriters) {
                        parser.text?.trim()?.let {
                            if (it.isNotEmpty()) songwriters.add(it)
                        }
                    }
                }
                XmlPullParser.END_TAG -> {
                    if (parser.name == "songwriter") {
                        inSongwriters = false
                    }
                }
            }
            eventType = parser.next()
        }

        // Create a special line at the end for songwriter credits.
        if (songwriters.isNotEmpty()) {
            val lastLineEnd = lines.maxOfOrNull { it.endMs } ?: 0L
            val text = "Written by ${songwriters.joinToString(", ")}"
            val tokens = text.split(" ")
            
            val words = mutableListOf<Word>()
            for (i in tokens.indices) {
                words.add(Word(tokens[i], lastLineEnd, lastLineEnd + 1000L, isPartOfWord = false))
            }
            
            lines.add(
                Line(
                    words = words,
                    startMs = lastLineEnd,
                    isSongwriter = true
                )
            )
        }

        // Inject instrumental interlude placeholders for significant gaps (>= 3s) between main lines.
        val mainLines = lines.filter { !it.isSongwriter }
            .sortedBy { it.startMs }
        val interludes = mutableListOf<Line>()
        for (i in 0 until mainLines.size - 1) {
            val gapStart = mainLines[i].endMs
            val gapEnd = mainLines[i + 1].startMs
            if (gapEnd - gapStart >= 3000L) {
                interludes.add(Line(
                    words = emptyList(),
                    startMs = gapStart,
                    interludeEndMs = gapEnd,
                    isInterlude = true,
                ))
            }
        }
        // Handle initial gap before the first line.
        if (mainLines.isNotEmpty() && mainLines.first().startMs >= 3000L) {
            interludes.add(Line(
                words = emptyList(),
                startMs = 0L,
                interludeEndMs = mainLines.first().startMs,
                isInterlude = true,
            ))
        }
        lines.addAll(interludes)
        lines.sortBy { it.startMs }

        return ParsedLyrics(lines, songwriters)
    }

    /**
     * Parses a <p> tag and its contents into a list of [Line] objects.
     * This handles nested <span> tags for background vocals and word-level timing.
     */
    private fun parseParagraph(
        parser: XmlPullParser,
        agent: String?,
        defaultAgent: String?,
    ): List<Line> {
        val pBegin = parser.getAttributeValue(null, "begin")?.let { parseTimeMs(it) } ?: 0L
        val pEnd = parser.getAttributeValue(null, "end")?.let { parseTimeMs(it) } ?: 0L

        val leadWords = mutableListOf<Word>()
        val backgroundGroups = mutableListOf<MutableList<Word>>()

        // Stack to track nested span timings and metadata.
        val stack = ArrayDeque<SpanContext>()
        stack.addLast(SpanContext(pBegin, pEnd, false))

        var previousEndedMidWord = false
        var currentBgGroup: MutableList<Word>? = null
        var inBgSpan = false
        var bgPreviousEndedMidWord = false

        var eventType = parser.next()
        while (!(eventType == XmlPullParser.END_TAG && parser.name == "p")) {
            when (eventType) {
                XmlPullParser.START_TAG -> {
                    // Inherit timing from parent if not specified.
                    val b = parser.getAttributeValue(null, "begin")?.let { parseTimeMs(it) }
                        ?: stack.last().begin
                    val e = parser.getAttributeValue(null, "end")?.let { parseTimeMs(it) }
                        ?: stack.last().end

                    // Check if this span represents background vocals.
                    val role = parser.getAttributeValue(null, "role")
                        ?: parser.getAttributeValue("http://www.w3.org/ns/ttml#metadata", "role")
                    val isBg = role == "x-bg" || stack.last().isBg

                    if (isBg && currentBgGroup == null) {
                        currentBgGroup = mutableListOf()
                        bgPreviousEndedMidWord = false
                    }
                    inBgSpan = isBg

                    stack.addLast(SpanContext(b, e, isBg))
                }
                XmlPullParser.TEXT -> {
                    val rawText = parser.text
                    val ctx = stack.last()
                    val isBgToken = ctx.isBg || inBgSpan

                    if (rawText.isNullOrBlank()) {
                        if (isBgToken) bgPreviousEndedMidWord = false else previousEndedMidWord = false
                    } else {
                        val startsWithSpace = rawText.first().isWhitespace()
                        val endsWithSpace = rawText.last().isWhitespace()
                        var trimmed = rawText.trim()

                        // Background tokens typically have parentheses, which we strip for cleaner UI.
                        if (isBgToken) {
                            trimmed = trimmed.removePrefix("(").removeSuffix(")")
                        }

                        if (trimmed.isNotEmpty()) {
                            // Split into words and sub-tokens (syllables).
                            val spaceTokens = trimmed.split("\\s+".toRegex()).filter { it.isNotEmpty() }
                            val wordsToAdd = mutableListOf<Pair<String, Boolean>>()
                            
                            for (i in spaceTokens.indices) {
                                val st = spaceTokens[i]
                                val subTokens = st.split(Regex("(?<=-)")).filter { it.isNotEmpty() }
                                for (j in subTokens.indices) {
                                    val isAttached = if (i == 0 && j == 0) {
                                        if (isBgToken) bgPreviousEndedMidWord && !startsWithSpace else previousEndedMidWord && !startsWithSpace
                                    } else {
                                        j > 0
                                    }
                                    wordsToAdd.add(Pair(subTokens[j], isAttached))
                                }
                            }
                            
                            val start = ctx.begin ?: pBegin
                            val end = ctx.end ?: (start + 1000L)
                            val duration = (end - start).coerceAtLeast(0)
                            // Distribute total span duration equally across all tokens within it.
                            val chunkDuration = if (wordsToAdd.isNotEmpty()) duration / wordsToAdd.size else duration

                            wordsToAdd.forEachIndexed { index, pair ->
                                val (token, isAttached) = pair
                                val wordStart = start + (index * chunkDuration)
                                val wordEnd = start + ((index + 1) * chunkDuration)
                                val wordDuration = wordEnd - wordStart

                                // Decide if this word should have granular letter-level animation.
                                val isLetterGroup = wordDuration >= 1000L && token.length <= 12 && token.length > 1
                                val letters = if (isLetterGroup) {
                                    val letterDuration = wordDuration.toFloat() / token.length
                                    token.mapIndexed { li, ch ->
                                        Letter(
                                            char = ch.toString(),
                                            startMs = wordStart + (li * letterDuration).toLong(),
                                            endMs = wordStart + ((li + 1) * letterDuration).toLong(),
                                        )
                                    }
                                } else emptyList()

                                val word = Word(token, wordStart, wordEnd, isPartOfWord = isAttached, isLetterGroup = isLetterGroup, letters = letters)
                                if (isBgToken) {
                                    currentBgGroup?.add(word)
                                } else {
                                    leadWords.add(word)
                                }
                            }
                            if (isBgToken) bgPreviousEndedMidWord = !endsWithSpace else previousEndedMidWord = !endsWithSpace
                        }
                    }
                }
                XmlPullParser.END_TAG -> {
                    // Close the current span context.
                    if (stack.isNotEmpty()) {
                        val popped = stack.removeLast()
                        if (popped.isBg && (stack.isEmpty() || !stack.last().isBg)) {
                            // If we finished a background span block, save the group.
                            currentBgGroup?.let { if (it.isNotEmpty()) backgroundGroups.add(it) }
                            currentBgGroup = null
                            inBgSpan = stack.isNotEmpty() && stack.last().isBg
                        }
                    }
                }
            }
            eventType = parser.next()
        }

        // Compare agent to default agent to determine alignment.
        val isOppositeAligned = agent != null && defaultAgent != null && agent != defaultAgent

        val result = mutableListOf<Line>()
        // Add the primary lead line.
        result.add(Line(leadWords, pBegin, agent = agent, isBackground = false, oppositeAligned = isOppositeAligned))
        // Add any associated background lines.
        for (bgGroup in backgroundGroups) {
            val bgStart = bgGroup.firstOrNull()?.startMs ?: pBegin
            result.add(Line(bgGroup, bgStart, agent = agent, isBackground = true, oppositeAligned = isOppositeAligned))
        }

        return result
    }

    /**
     * Parses a TTML time format string into milliseconds.
     * Supports mm:ss.ms and hh:mm:ss.ms formats.
     */
    private fun parseTimeMs(time: String): Long {
        val parts = time.split(":")
        if (parts.size == 2) {
            val min = parts[0].toLong()
            val secParts = parts[1].split(".")
            val sec = secParts[0].toLong()
            val ms = if (secParts.size > 1) secParts[1].padEnd(3, '0').take(3).toLong() else 0L
            return (min * 60 + sec) * 1000 + ms
        } else if (parts.size == 3) {
            val hrs = parts[0].toLong()
            val min = parts[1].toLong()
            val secParts = parts[2].split(".")
            val sec = secParts[0].toLong()
            val ms = if (secParts.size > 1) secParts[1].padEnd(3, '0').take(3).toLong() else 0L
            return (hrs * 3600 + min * 60 + sec) * 1000 + ms
        }
        return 0L
    }
}
